google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITE2ETracingTest.java

```diff
@@ -0,0 +1,1251 @@
+    fetchAndValidateNonTransactionTrace(
+        customSpanContext.getTraceId(),
+        SPAN_NAME_DOC_REF_LIST_COLLECTIONS,
+        grpcSpanName(LIST_COLLECTIONS_RPC_NAME));
+  }
+
+  @Test
+  public void getAllTraceTest() throws Exception {
+    // Make sure the test has a new SpanContext (and TraceId for injection)
+    assertNotNull(customSpanContext);
+
+    // Inject new trace ID
+    Span rootSpan = getNewRootSpanWithContext();
+    try (Scope ss = rootSpan.makeCurrent()) {
+      DocumentReference docRef0 = firestore.collection("col").document();
+      DocumentReference docRef1 = firestore.collection("col").document();
+      DocumentReference[] docs = {docRef0, docRef1};
+      firestore.getAll(docs).get();
+    } finally {
+      rootSpan.end();
+    }
+    waitForTracesToComplete();
+
+    fetchAndValidateNonTransactionTrace(
+        customSpanContext.getTraceId(), grpcSpanName(BATCH_GET_DOCUMENTS_RPC_NAME));
+  }
+
+  @Test
+  public void queryGetTraceTest() throws Exception {
+    // Make sure the test has a new SpanContext (and TraceId for injection)
+    assertNotNull(customSpanContext);
+
+    // Inject new trace ID
+    Span rootSpan = getNewRootSpanWithContext();
+    try (Scope ss = rootSpan.makeCurrent()) {
+      firestore.collection("col").whereEqualTo("foo", "my_non_existent_value").get().get();
+    } finally {
+      rootSpan.end();
+    }
+    waitForTracesToComplete();
+
+    fetchAndValidateNonTransactionTrace(
+        customSpanContext.getTraceId(), SPAN_NAME_QUERY_GET, grpcSpanName(RUN_QUERY_RPC_NAME));
+  }
+
+  @Test
+  public void transactionTraceTest() throws Exception {
+    // Make sure the test has a new SpanContext (and TraceId for injection)
+    assertNotNull(customSpanContext);
+
+    // Inject new trace ID
+    Span rootSpan = getNewRootSpanWithContext();
+    try (Scope ss = rootSpan.makeCurrent()) {
+      firestore
+          .runTransaction(
+              transaction -> {
+                Query q = firestore.collection("col").whereGreaterThan("bla", "");
+                DocumentReference d = firestore.collection("col").document("foo");
+                DocumentReference[] docList = {d, d};
+                // Document Query.
+                transaction.get(q).get();
+
+                // Aggregation Query.
+                transaction.get(q.count());
+
+                // Get multiple documents.
+                transaction.getAll(d, d).get();
+
+                // Commit 2 documents.
+                transaction.set(
+                    firestore.collection("foo").document("bar"),
+                    Collections.singletonMap("foo", "bar"));
+                transaction.set(
+                    firestore.collection("foo").document("bar2"),
+                    Collections.singletonMap("foo2", "bar2"));
+                return 0;
+              })
+          .get();
+    } finally {
+      rootSpan.end();
+    }
+    waitForTracesToComplete();
+
+    fetchAndValidateTransactionTrace(
```

Comment:
> each invocation of fetchAndValidateTransactionTrace results in at least one traceClient_v1.getTrace

The actual issue is that `getTrace` may fetch an `incomplete` trace due to the eventual consistency of trace ingestion. So, we need to retry for 2 cases - for `trace not found` and for `incomplete trace`. So multiple calls to `getTrace` are unavoidable to fight flakiness.

> Then you can merge fetchAndValidateNonTransactionTrace and fetchAndValidateTransactionTrace

For non-transaction traces, the compareTo callstack is exhaustive i.e. it's the entire trace structure so we can check for number of spans in the passed-in compareTo callstack. 

For transaction traces that *may not* be true and the expected number of spans may be different from those of one callstack. To keep the methods as simple as possible and obvious for the reader, I chose to split the two.

Reply:
I understand the need to call `getTrace` more than once until the trace is fully available in Cloud Trace.

but let's assume that the trace is already available in Cloud Trace, we are still calling `fetchAndValidateTransactionTrace` 5 times, and so effectively calling `getTrace` 5 times and succeeding on the first try.

Isn't it true that by making 1 call:

```java
fetchAndValidateTransactionTrace(
  Arrays.asList(
    Arrays.asList(SPAN_NAME_TRANSACTION_RUN, SPAN_NAME_TRANSACTION_BEGIN, grpcSpanName(BEGIN_TRANSACTION_RPC_NAME)),
    Arrays.asList(SPAN_NAME_TRANSACTION_RUN, SPAN_NAME_TRANSACTION_GET_QUERY, grpcSpanName(RUN_QUERY_RPC_NAME)),
    Arrays.asList(SPAN_NAME_TRANSACTION_RUN, SPAN_NAME_TRANSACTION_GET_AGGREGATION_QUERY, grpcSpanName(RUN_AGGREGATION_QUERY_RPC_NAME)),
    Arrays.asList(SPAN_NAME_TRANSACTION_RUN, SPAN_NAME_TRANSACTION_GET_DOCUMENTS, grpcSpanName(BATCH_GET_DOCUMENTS_RPC_NAME)),
    Arrays.asList(SPAN_NAME_TRANSACTION_RUN, SPAN_NAME_TRANSACTION_COMMIT, grpcSpanName(COMMIT_RPC_NAME))
  );
```
you don't need the `numExpectedSpans` anymore?

(inside `fetchAndValidateTransactionTrace` you can make a `Set` and the size of the set is the expected number of spans).

And you can retry until you have the expected number of spans. Once you do, you can call `containsCallStack` once for each stack trace. And most importantly once you do this I think you can merge the two transaction and non-transaction functions. Since this is in test code, the simplification that results is more important than the performance.

LMK if I'm missing something? happy to discuss offline as well.

Reply:
Ah I see your point. You're right. This can be optimized to just call 1 `getTrace` - I'll fix it.